Lib/unittest/mock.py

```diff
@@ -509,7 +514,11 @@ def _mock_add_spec(self, spec, spec_set, _spec_as_instance=False,
                                         _spec_as_instance, _eat_self)
             _spec_signature = res and res[1]
 
-            spec = dir(spec)
+            spec = []
+            for member_k, member_v in _members:
+                spec.append(member_k)
+                if member_k == '__init__' and hasattr(member_v, '__code__'):
```

Comment:
I am not experienced with code object. This also seems to end up including the functions and methods as part of `__init__` as per breakpoint here below where startswith is represented as a valid attribute. In case you are trying to get signature of the `__init__` you can use inspect module : https://docs.python.org/3/library/inspect.html#introspecting-callables-with-the-signature-object . But not all the methods in signature become attributes as they could be flags to set another set of attributes dynamically too.

```python
class Person:

    def __init__(self, name):
        self.name = name
        foo = 1

        if name.startswith("Test"):
            print("Test user")


m_instance = Mock(spec=Person("John"))
print(m_instance.name)
print(m_instance.startswith)
```

```
./python /tmp/foo.py
> /home/karthikeyan/stuff/python/cpython/Lib/unittest/mock.py(522)_mock_add_spec()
-> spec.extend(member_v.__code__.co_names)
(Pdb) member_v.__code__.co_names
('name', 'startswith', 'print')
(Pdb) c
<Mock name='mock.name' id='140293864592656'>
<Mock name='mock.startswith' id='140293862696912'>
```

Reply:
Thanks for the feedback. Those are good points. When I have some free time I will see if I can find a more surgical solution and and in some test cases for the examples you folks gave.